Add Ant 03

Smittytron opened the pull request:
This mission may need some stress testing for difficulty. Ant01 and Ant04 have mostly infantry vs ants while 02 and 03 have Tanks. Tanks hit moving targets in OpenRA (not to mention fire on the move), which makes the original ant stats weak in this mission, but strong in the first and fourth one.

I added a commit to give the anthill the correct impassable tiles.

Smittytron committed:
Change anthill to correct impassable tiles

Smittytron committed:
Add Ant 03

darcyrush commented:
Would you mind rebasing this please since make has been updated since this was submitted

darcyrush commented:
Yes, hard isn't a challenge with tanks and turrets, never mind the APC rushing. Perhaps nerfing tank shell damage against them would keep their difficultly in the later missions similar to how they are in the first infantry based mission. And to match the original film they are based on, perhaps make their weakness fire (as well as grenades). I believe they are no longer buildable via the biolab in multiplayer so balance considerations should be able to be discussed?

Some general quality of life changes for this mission could include reducing the scientist selection priority, as currently they are the same as infantry and tanks. And also when the nests have been gassed, it would be nice if the remaining ants just attacked the base. 

I also noticed the ants never take the flanking route on the right when attacking from the north and rarely take the southern mud bridge when attacking from the west which makes defending pretty boring. 

Balance aside, the map plays as expected.

Smittytron commented:
Updated. Added waypoints to vary up the attack paths and added a function to send the remaining surface ants once the nests have all been gassed. I ended up doubling the ant HP here. It's not as clean as if we had the same stats across all four missions but considering the missions are about fighting giant ants I think we can get away with it.

tttppp commented:
I tried all three difficulty levels. I completed Easy and Medium but couldn't get past gassing the first nest on Hard.  Generally seems good to me. Some observations.

Ants come out of nests and die after. I assume this is deliberate, I didn't check what happened in the original.  This has the weird side effect of the 'dead' ants tanking against my units but also when the ants are blocked then they're able to continue to attack my units indefinitely.

Sometimes streams of ants walk into other nests and then die. This seems fine - it's thematically correct at least.

At the beginning I got two APCs, but these don't allow me to use the "select all" shortcut to select them both.

Smittytron commented:
Added a selectable class to solve the APC issue. 

> Ants come out of nests and die after. I assume this is deliberate

Yes, this is what the original did.

tttppp commented:
> Added a selectable class to solve the APC issue.

Verified that this is now fixed. Thanks!

tttppp commented on mods/ra/maps/ant-03/ant-03-AI.lua:
```diff
@@ -0,0 +1,104 @@
+
+ActivateAntHives = function()
+	local difficulty = Map.LobbyOption("difficulty")
+	AntSquad = AntSquad[difficulty]
```
Is it normal to reuse global variables like this?  It would seem more natural to have two separate variables for the squad for this game and the table of all possible squads.

Smittytron replied:
It's consistent with what we do with difficulty levels in general (check Allies08 for example), but I'm not sure that's the cleanest way overall.

tttppp commented on mods/ra/maps/ant-03/ant-03.lua:
```diff
@@ -0,0 +1,199 @@
+				end
+			end)
+		end
+	end)
```
There's quite a lot of repeated code here. I think you could use arrays to reduce it to a loop:
```
for nestId = 1, 7 do
	Trigger.OnEnteredProximityTrigger(Hives[nestId].CenterPosition, WDist.FromCells(1), function(actor, id)
		if actor.Type == "chan" then
			Trigger.RemoveProximityTrigger(id)
			HiveGassed[nestId] = true
			Actor.Create("flare", true, { Owner = England, Location = Hives[nestId].Location })
			Actor.Create("flare", true, { Owner = England, Location = HiveFlares[nestId].Location })
			Trigger.OnEnteredFootprint(HiveKillzones[nestId], function(ant)
				if ant.Type == "warriorant" then
					ant.Kill("ExplosionDeath")
				end
			end)
		end
	end)
end
```

tttppp commented:
I added a couple of comments on the Lua code, but generally looked fine.  Fun level - thanks!

Smittytron commented:
Updated

tttppp reviewed (approved):
LGTM

reaperrr reviewed (approved):
We can still fine-tune difficulty of the ant campaign later, LGTM.

reaperrr commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/29160b3df3171ba7b55b9687ddb986581c885371...bd5c357abcd3fa577f2e31914737a7d59321b69f
